pkg/shell/plot.js

```diff
@@ -310,263 +287,63 @@ shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
-                var val = sample_real_time(t - interval, t);
-                if (val !== null)
-                    series_data[series_data.length] = [ t, val ];
-            }
-        }
-
-        reset_series();
-        add_series();
-
-        function process_samples(vals) {
-            real_time_samples_pos += 1;
-            if (real_time_samples_pos >= max_real_time_samples)
-                real_time_samples_pos = 0;
-            else
-                n_real_time_samples += 1;
-            real_time_samples[real_time_samples_pos] = vals[0][1];
-            real_time_samples_timestamp = vals[0][0];
+            return val;
         }
 
-        function channel_closed(ev, options, desc) {
-            real_time_channel = null;
-            if (options.problem && options.problem != "terminated" && options.problem != "disconnected")
-                console.log("problem retrieving " + desc + " metrics data: " + options.problem);
-        }
+        function reset_series() {
+            if (channel)
+                channel.close();
 
-        real_time_channel = channel_sampler($.extend({ }, chanopts, {
-            source: "direct",
-            interval: real_time_samples_interval
-        }), process_samples);
-        $(real_time_channel).on("close", function (event, options) {
-
-            /* Go for the fallback internal metrics if these metrics are not supported */
-            if ((options.problem == "not-supported" || options.problem == "not-found") && fallback) {
-                real_time_channel = channel_sampler($.extend({ }, chanopts, {
-                    source: "internal",
-                    interval: real_time_samples_interval,
-                    metrics: fallback
-                }), process_samples);
-                $(real_time_channel).on("close", function(event, options) {
-                    channel_closed(event, options, "internal");
-                });
-
-            /* Otherwise it could be just a normal failure */
-            } else {
-                channel_closed(event, options, "real time");
-            }
-        });
+            channel = cockpit.metrics(interval, chanopts);
 
-        function sample_real_time(t1, t2) {
-            var p1 = Math.min(Math.floor((t1 - real_time_samples_timestamp) / real_time_samples_interval), 0);
-            var p2 = Math.min(Math.ceil((t2 - real_time_samples_timestamp) / real_time_samples_interval), 0);
+            var metrics_row = grid.add(channel, [ ]);
+            var factor = desc.factor || 1;
+            series.data = grid.add(function(row, x, n) {
+                for (var i = 0; i < n; i++)
+                    row[x + i] = [(grid.beg + x + i)*interval, flat_sum(metrics_row[x + i])*factor];
+            });
 
-            if (-p1 >= n_real_time_samples)
-                p1 = -n_real_time_samples + 1;
 
-            if (p1 > p2)
-                return null;
+            channel.follow();
+            channel.fetch(grid.beg, grid.end);
```

Comment:
The cockpit.metrics did not implement the necessary hook. I have done that now, and added the 'for_walking' flag that we discussed on irc.  Scrolling now appears smoother, but after a few scrolls, things start to go wrong and both the browser and cockpit-pcp do excessive work for a long time.